### DISCUSSION TITLE
是否可以创造“国内被动连接”的通道？

### DISCUSSION THREAD

**Discussion by Kylejustknows** [v2fly/v2ray-core#1432] [Ideas]
因为防火墙的“解密压力”巨大，测试发现，墙只对 “从国内往外连接” 进行各种嗅探和解密，对 “从国外往内的连接” 管理松懈。

判断：
我有两台服务器，一台国内一台国外，

国外机作为客户端，主动连接国内v2服务器（看国内版权视频音乐等），非加密协议都畅通。
国内机作为客户端，连接国外，各种断流、被封端口，

希望V2ray能实现一种，国外服务器（在配置后），主动连接国内某电脑的IP和端口，“被动建立长联通道” 后继续数据传输。

最好的结果：实现稳定上网； 最坏的结果：让防火墙的运算量2x加倍。

（现在通过一些复杂的方法勉强成功了，但非常非常麻烦。搜了一下，这个问题也有别人考虑过 https://community.geph.io/t/topic/3983 ）

**Comment by Kylejustknows** [comment]
简述 “复杂的方法”：

1，从国外v2连回国内 （反向连接非常稳定）

2，国内服务器上，安装随便普通的vpn服务端

3，国外机器装vpn客户端，这个vpn客户端强制使用代理（走第1步建立的v2网络）连到国内vpn服务器 （要注意局域网地址的可能block要调整，如果走代理成功的话，连接127.0.0.1就是连接国内服务器自己，如果走代理失败，连接127.0.0.1会连到国外服务器自己）

4，国内服务器，也装vpn客户端自己连接到自己。

（这时，国内国外服务器上都各有一个虚拟网卡，都有独立的虚拟IP地址，互相能ping通，这两个IP地址之间的所有流量，都被包裹在VPN里，而VPN又被包裹在第1步建立的V2ray的反向通道里）

(2~4步，还有一个可能稍微简单一点的做法：Teamviewer 强行使用IP地址（而不是ID）双机直连，开启Teamviewer 的VPN，直接建立两台机子之间的点对点VPN) 

5, 海外机子，V2开启socks代理, 要点： a)这个代理强行routing到freedom, b)这个代理只监听从VPN虚拟网卡地址建立的连接。

6，这时，国内的机子就已经能 直接使用Socks代理 （连国外机子时使用VPN虚拟的IP地址）来上网了。 所有流量都被包裹在VPN里，VPN又被包裹在V2Ray里。 当然国内机子也可以再开一个本地Socks方便国内其他电脑调用。

**Reply by krrr** [reply]
用vpn太复杂了，用你给的链接里提到的frp会简单些

**Reply by Kylejustknows** [reply]
非常谢谢啊，是个神软件！frp应该是能够代替复杂的VPN实现反向连接。
我还没有真正使用，但是看了文档，能够“强行指定使用本地代理”，所以反向连接应该是能成功的。
就是，它好像UDP和TCP必须单独设置，到两个端口分开的映射，Socks会有点麻烦，但其它vmess等协议都是UDP over TCP的，所以问题不大，能解决。

我暂时还是开着VPN，因为两台机器都有局域网IP，非常方便，我好做一些稀奇古怪的事情和测试。

**Reply by krrr** [reply]
我照你说的做了，连接非常稳定，效果卓越！不像之前我国内主动连国外时，会发生每隔十几分钟到几小时被阻断几分钟的情况（普通看网页应该影响不大，我是打外服游戏用，非常致命游戏会掉线。很奇怪gfw为什么是这种阻断方式）

我现在的配置：
1、frp从国外向国内发起连接，把外网一个v2ray服务端口映射到国内服务器上的一个端口（frp开启加密）
2、v2ray客户端直接连国内服务器公网ip暴露出来的v2ray端口（v2ray流量加密，防止云服务厂商检测）

**Reply by Kylejustknows** [reply]
嗯嗯，有点意思哈哈。 

你的方法，是把v2的数据包裹在frp的加密流量内。  
我的方法，是把frp的数据包裹在v2的加密流量里。

唯一的区别是，因为frp没有针对“反监控”进行设计，你的方法，中间人通过一些嗅探后能知道 “你在用frp做一些事情”，但估计防火墙也不在意，协议太小众了，你告诉我之前，我都不知到有frp这样的东西。

因为frp有数据加密，所以用你这种方法“反向翻墙”，连v2ray都可以省了（因为流量加密了，具体数据内部传递的是vmess数据包还是别的什么都没有区别了，中间人无法探测），随便开个代理服务器，端口直接frp反向连接到家里路由器就全家上网了。

甚至如上面你的方法中，直接反向到某个国内云服务器上，不用v2都没事，你多虑了。因为方法中唯一暴露的，是国内的两台电脑的之间互联数据，而国内两台电脑之间的互联的数据压根没人查，不管你的协议是socks还是L2TP （商业云服务器上大半都装都有某种VPN来内网同步后台数据等等，VPN在公司远程办公和数据访问也被广泛使用。只要不VPN到海外，云厂商或ISP都压根没人会管你）

我的担心是，你的方法，如果被广泛使用来翻墙，很快frp会被防火墙加入"黑名单协议"封掉，导致很多使用frp来做正常网络服务的用户受到影响。

所以我希望，看到此讨论受到启发的用户，尽量使用我的方法：先用v2ray反向连接到国内后，再在v2ray的流量内，开启vpn或者frp等来方便使用，更隐蔽更安全一些，最终效果一样，

**Reply by krrr** [reply]
我觉得你想太多了，frp绝大部分的使用场景是frp架设在国内云上，然后内网设备去连。反而需要担心的点是，对于这种反向翻墙不会触发一些高级的深度包检测的现状，会随着反向翻墙的人变多而被针对，就像你给的链接里面有的人提出的一样

其实我省略了一些无关紧要的细节，比如我在frp外面套了kcptun，加密的是最外层而不是frp本身。为了游戏流畅进行了加速……我直接用v2ray是因为它有透明代理功能，使用起来还是比单纯的socks5服务器方便。另外我没有公网ip，所以才在国内云上起frp服务器。这种方式也注定不可能广泛使用，现在绝大部分家宽都没有公网ip地址了，就算有也要搞ddns麻烦死了

我在想，就算以后反向翻墙真的也被针对，至少也增加了gfw的压力，恶心了它们一把，值了。以后的翻墙协议就不能表现得完全随机了，得尽量藏在常见协议里，例如v2的裸websocket模式一样

**Reply by zzlinwq** [reply]
> 把外网一个v2ray服务端口映射到国内服务器上的一个端口

是否是FRP常用的端口7000？可以具体说说吗？

**Comment by kingwilliam** [comment]
长话短说

不行 墙会观察国外流入的数据 不会理会由谁发起连接

您在国外翻入国内看如iqiyi, 都是http plain text数据流出 看得没那麽紧

(当然还没那麽简单)

**Reply by Kylejustknows** [reply]
> 长话短说
> 
> 不行 墙会观察国外流入的数据 不会理会由谁发起连接
> 
> 您在国外翻入国内看如iqiyi, 都是http plain text数据流出 看得没那麽紧
> 
> (当然还没那麽简单)


如果您有任何建设防火墙的经验，你会知道 “传入连接” 和 “传出连接” 这连接方向是最重要的参数之一，在每一层防火墙上都是非常区别对待的。


我也长话短说一下哈哈哈哈：

真实测试！（这两天刚刚做的）国内国外两台主机（一台上海一台新加坡），同样的V2协议QUIC，连接后做了几小时差不多同样的事情来测试：

a), 国内机做v2客户端，国外机v2服务器 （所有连接都是“出站连接”，国内机对着国外机的一个v2端口发包）, 网络明显的延迟，随机突然爆发式丢包几次。

b) 国外机做客户端，国内开v2服务器 （所有连接都是入站连接，国外机对着国内的一个v2端口猛发包），网速瞬间跑满，全程流畅无任何丢包。

以上测试，都是先连通v2，再在v2的代理下建立新的一层p2p双向vpn，然后在vpn的两个虚拟IP下再互开一个socks代理，然后做了同样的测试：看外网视频、看国内视频等等，出国的大量数据，回国的大量数据等等都测试了很久。


我的猜测：

现在防火墙的嗅探和解密即时分析每秒几万G的数据量，压力非常大，已经几乎跑满了设备负荷（以至于经常处理不过来导致丢包），为了降低负荷和减少访问国内服务的误伤，“高级的审查和嗅探功能” 和只对国内发起握手的TCP或UDP包进行过滤。

国内主动握手发起的链接，每个字节都进行过滤。  

而国外发起、主动连接国内的链接，只进行简单的协议判断后就放行数据包。


> 补充:
> 
>     1. 国内外vps中间已用v2, 看不出vpn有什麽加持
> 
>     2. 在国内vps安装v2就要 非常非常 千万千万 小心.
> 
>     3. 开长链非常容易给墙注意

1）国内外vps中间已用v2，但v2默认是单向的 （你连接某个国外v2服务器后，你能访问国外网络，但国外的那台机子无法通过这条v2连接，来反向访问你国内本地的东西）。 
加了一层VPN后，内外网络就成了桥连，能互相访问。  我做这个目的，就是为了反着来： A主动握手连接B，在连接建立后，让B访问A的网络资源。

2）谢谢提醒。虽然用的是xx云，但用windows主机，关掉了云监控（只留下纯windows），账户强密码加密、文件也手动EFS加密后，锁屏。 
（除非xx云dump我的全部windows内存然后请安全高手做内存dump分析，否则无法发现我在运行什么。我觉得我个人看个视频等家庭使用，达不到这么高的关注程度哈哈）
不过您的提醒，对新手还是很重要的。

3）网络最外层使用的就是普通的v2，（持续连接的vpn都是跑在v2流量的包裹之下），所以从网络上没有什么长连接。

**Comment by kingwilliam** [comment]
补充:
1. 国内外vps中间已用v2, 看不出vpn有什麽加持
2. 在国内vps安装v2就要 非常非常 千万千万 小心.
3. 开长链非常容易给墙注意

**Comment by xxx-zzl** [comment]
vpn能被包裹在V2ray的通道里吗？vpn应该比v2更底层

**Reply by krrr** [reply]
当然可以啊，很多vpn协议基于tcp/udp，但是性能会很差建议换掉v2ray

**Comment by hahagugu233** [comment]
我个人认为这么做是有显著特征的，境外主动连接家宽算什么事，你用家宽连国外域名，GFW还能理解为你在看网站，你这一搞不就显著特征出来了？
而且普及率和可操作性不是很高，首先是公网IP，当然也可以用FRP，不过如果使用frp，那么如何保证速度呢？大多数frp都是限速的，这不是增加麻烦吗？

**Comment by zhpjy** [comment]
尝试了有似乎有效果。不需要frp，可以直接使用 v2ray 自带的反代来实现。具体配置如下
国外机器
```
{
  "log": {
    "access": "",
    "error": "",
    "loglevel": "warning"
  },
  "reverse": {
    "bridges": [
      {
        "tag": "bridge",
        "domain": "allbridge.v2ray.com"
      }
    ]
  },
  "outbounds": [
    {
      "tag": "bridgeout",
      "protocol": "freedom"
    },
    {
      "protocol": "vmess",
      "settings": {
        "vnext": [
          {
            "address": "国内服务器的公网IP",
            "port": 60210,
            "users": [
              {
                "id": "134b53ca-b0cc-44a7-a28f-4214842c2fd6"
              }
            ]
          }
        ]
      },
      "tag": "interconn"
    }
  ],
  "routing": {
    "rules": [
      {
        "type": "field",
        "inboundTag": [
          "bridge"
        ],
        "domain": [
          "full:allbridge.v2ray.com"
        ],
        "outboundTag": "interconn"
      },
      {
        "type": "field",
        "inboundTag": [
          "bridge"
        ],
        "outboundTag": "bridgeout"
      }
    ]
  }
}

```
国内机器
```
{
  "log": {
    "access": "",
    "error": "",
    "loglevel": "warning"
  },
  "reverse": {
    "portals": [
      {
        "tag": "portal",
        "domain": "allbridge.v2ray.com"
      }
    ]
  },
  "inbounds": [
    {
      "tag": "portalin",
      "port": 60211,
      "protocol": "vmess",  // 用于client连接
      "settings": {
        "clients": [
          {
            "id": "89682891-3d57-4cef-abbb-fbac5937ba29"
          }
        ]
      }
    },
    {
      "port": 60210,
      "tag": "interconn",
      "protocol": "vmess",
      "settings": {
        "clients": [
          {
            "id": "134b53ca-b0cc-44a7-a28f-4214842c2fd6"
          }
        ]
      }
    }
  ],
  "routing": {
    "rules": [
      {
        "type": "field",
        "inboundTag": [
          "portalin"
        ],
        "outboundTag": "portal"
      },
      {
        "type": "field",
        "inboundTag": [
          "interconn"
        ],
        "outboundTag": "portal"
      }
    ]
  }
}

```
需要魔法的客户端使用国内机器的60211端口即可

**Reply by zzlinwq** [reply]
allbridge.v2ray.com 这个域名指向哪里？起到什么作用呢？

**Reply by zhpjy** [reply]
> allbridge.v2ray.com 这个域名指向哪里？起到什么作用呢？
可以了解下 v2ray 的反向代理配置。这个是用来标识用的。和具体的域名没有任何关系。

**Comment by b23r0** [comment]
可以考虑测测反向socks5和正向socks5在gfw是否有明显差异。

po一个我写的 https://github.com/b23r0/rsocx

**Comment by aa51513** [comment]
我这样子用着好久了，直接frp的socks5到路由器的内网，v2ray都没有用，稳的一匹

**Reply by zzlinwq** [reply]
具体你是怎么实现的？用你楼上的办法？

**Reply by zzlinwq** [reply]
frp服务端在国内？

**Reply by agnescxt** [reply]
问下这种情况下登陆国外网站的ip是显示的国外的ip吗？

**Reply by BanKnight** [reply]
完全不加一点加密吗？还是蛮大胆的。
再就是这种方案，应该不支持udp，要玩游戏的不太适合了

**Reply by garryforreg** [reply]
frp本身就是加密的,socks5的流量是包装在frp的通道里的,目前用这种方式很久了,我是用的国外朋友的家庭宽带,因为朋友家宽没有公网ip所以用这个方法借用朋友的家宽使用。

**Reply by RookieTerry** [reply]
> frp本身就是加密的,socks5的流量是包装在frp的通道里的,目前用这种方式很久了,我是用的国外朋友的家庭宽带,因为朋友家宽没有公网ip所以用这个方法借用朋友的家宽使用。

我也打算这么弄，不过是在国外朋友家里搭建vmess服务，frp可以反向代理vmess的流量吗？

**Comment by CodingXiaoheng2** [comment]
https://linux.do/t/topic/290832
还可以搭配上WARP:)

**Comment by RookieTerry** [comment]
那可以实现在内网的设备上运行vmess服务，然后用frp反向代理vmess流量吗？vmess使用自签名证书开启了tls加密
